src/main/java/zoe/ui/Ui.java

```diff
@@ -0,0 +1,165 @@
+    private Image loadImage(String resourcePath) {
+        var imageStream = Objects.requireNonNull(this.getClass().getResourceAsStream(resourcePath));
+        return new Image(imageStream);
+    }
+
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    /**
+     * Creates two dialog boxes, one echoing user input and the other containing Duke's reply and then appends them to
+     * the dialog container. Clears the user input after processing.
+     */
+    @FXML
+    private void handleUserInput() {
+        String input = userInput.getText();
+        dialogContainer.getChildren().add(DialogBox.getUserDialog(input, userImage));
+        zoe.notify(input);
+        userInput.clear();
+    }
+
+    /**
+     * Displays a response from Zoe (the chatbot) in the dialog container.
+     *
+     * @param response The message to be displayed in Zoe's dialog box.
+     */
+    public void reply(String response) {
+        dialogContainer.getChildren().add(DialogBox.getDukeDialog(response, dukeImage));
+    }
+
+    /**
+     * Displays a welcome message when the chatbot starts.
+     */
+    public void showWelcome() {
+        reply("Hello! I'm Zoe\nWhat can I do for you?\n");
+    }
+
+    /**
+     * Displays an error message when an exception occurs.
+     *
+     * @param e The exception containing the error message to be displayed.
+     */
+    public void showError(ZoeException e) {
+        reply(e.getMessage());
+    }
+
+    /**
+     * Displays a confirmation message after a task is added to the task list.
+     *
+     * @param taskList The current task list.
+     * @param task The task that was added.
+     */
+    public void showAddedTask(TaskList taskList, Task task) {
+        reply(String.format("""
+                Got it. I've added this task:
+                \t%s
+                Now you have %d task(s) in your list.""", task.getFullDescription(), taskList.size()));
+    }
+
+    /**
+     * Displays a confirmation message after a task is removed from the task list.
+     *
+     * @param taskList The current task list.
+     * @param task The task that was removed.
+     */
+    public void showRemovedTask(TaskList taskList, Task task) {
+        reply(String.format("""
+                Noted. I've removed this task:
+                \t%s
+                Now you have %d task(s) in your list.""", task.getFullDescription(), taskList.size()));
+    }
+
+    /**
+     * Displays a list of tasks that match a search or filter criteria.
+     *
+     * @param tasks The list of tasks to be displayed.
+     */
+    public void showList(List<? extends Task> tasks) {
+        reply("Here are the matching tasks in your list:\n"
+                + IntStream
+                .range(0, tasks.size())
+                .mapToObj(i -> String.format("%d. %s", i + 1, tasks.get(i).getFullDescription()))
+                .collect(Collectors.joining("\n")));
+    }
```

Comment:
Maybe use multiline string literal instead of concatenation to have consistent formatting with the rest of the code, thereby making the code easier to read and maintain?
```suggestion
    public void showList(List<? extends Task> tasks) {
    String taskListStr = IntStream
            .range(0, tasks.size())
            .mapToObj(i -> String.format("%d. %s", i + 1, tasks.get(i).getFullDescription()))
            .collect(Collectors.joining("\n"));

    reply(String.format("""
            Here are the matching tasks in your list:
            %s""", taskListStr));
}
```